presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -585,5 +596,119 @@ public void prepareFinal()
+                if (rowBlockBuilder == null) {
+                    rowBlockBuilder = new RowBlockBuilder(spillingTypes, null, (int) groupIdsBlock.getGroupCount());
+                }
+
+                BlockBuilder output = rowBlockBuilder.beginBlockEntry();
+                for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                    spillingTypes.get(channel).appendTo(page.getBlock(channel), pos, output);
+                }
+                rowBlockBuilder.closeEntry();
+
+                intermediateState.set(groupId, rowBlockBuilder);
+            }
+
+            // TODO do this in prepareFinal() instead because this duplicates data in intermediateState and distincting hash table
+            delegate.addInput(groupIdsBlock, page);
+        }
+
+        @Override
+        public void addIntermediate(GroupByIdBlock groupIdsBlock, Block block)
+        {
+            ArrayBlock arrayBlock = (ArrayBlock) block;
+            int newPositionCount = arrayBlock.getRawElementBlock().getPositionCount();
+            long[] newGroupIds = new long[newPositionCount];
+            boolean[] nulls = new boolean[newPositionCount];
+            int currentRowBlockIndex = 0;
+            for (int groupIdPosition = 0; groupIdPosition < groupIdsBlock.getPositionCount(); groupIdPosition++) {
+                for (int unused = 0; unused < arrayBlock.getBlock(groupIdPosition).getPositionCount(); unused++) {
+                    newGroupIds[currentRowBlockIndex] = groupIdsBlock.getGroupId(groupIdPosition);
+                    nulls[currentRowBlockIndex] = groupIdsBlock.isNull(groupIdPosition);
+                    currentRowBlockIndex++;
+                }
+            }
+
+            Block[] blocks = new Block[spillingTypes.size()];
+            for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                blocks[channel] = ((RowBlock) arrayBlock.getRawElementBlock()).getRawFieldBlocks()[channel]; // super ugly exposing block internals
+            }
+            GroupByIdBlock squashedGroupIds = new GroupByIdBlock(groupIdsBlock.getGroupCount(), new LongArrayBlock(newPositionCount, Optional.of(nulls), newGroupIds));
+            delegate.addIntermediateStateOnUnspill(squashedGroupIds, blocks);
+        }
+
+        @Override
+        public void evaluateIntermediate(int groupId, BlockBuilder output)
+        {
+            BlockBuilder arrayBlockBuilder = output.beginBlockEntry();
+            RowBlock rowBlock = (RowBlock) intermediateState.get((long) groupId).build();
+            for (int i = 0; i < rowBlock.getPositionCount(); i++) {
+                arrayBlockBuilder.appendStructure(rowBlock.getBlock(i));
+            }
+            output.closeEntry();
+        }
+
+        @Override
+        public void evaluateFinal(int groupId, BlockBuilder output)
+        {
+            delegate.evaluateFinal(groupId, output);
+        }
+
+        @Override
+        public void prepareFinal()
+        {
```

Comment:
Add a check state to make sure `intermediateState` is empty. We shouldn't touch `intermediateState` at this step anymore.

Now, this is the key part. Let's take all the data in `unspilledBlocks` and feed them into `delegate.addInput(...)`. This should be the first time we touch `delegate`. Not that if `unspilledBlocks` is a list or an array, we should deref the item so that we don't use 2X the memory. The easiest way to do so is to assign null to the entry that has just been fed into `delegate.addInput(...)`.